libraries/mongodb.rb

```diff
@@ -168,38 +174,32 @@ def self.configure_replicaset(node, name, members)
         end
         Chef::Log.error("configuring replicaset returned: #{result.inspect}") unless result.fetch('errmsg', nil).nil?
       else
-        Chef::Log.info 'going to add and remove members from the replicaset'
-        # remove removed members from the replicaset and add the new ones
-        old_ids = config['members'].map { |member| member['_id'] }
-        rs_members.map! { |member| member['host'] }
-        config['version'] += 1
-        old_members = config['members'].map { |member| member['host'] }
-        members_delete = old_members - rs_members
-        config['members'] = config['members'].delete_if { |m| members_delete.include?(m['host']) }
-        config['members'].map! do |m|
-          host = m['host']
-          { '_id' => m['_id'], 'host' => host }.merge(rs_options[host])
-        end
+        Chef::Log.info 'going to update the members of the replicaset'
+        curr_members = config['members'].dup
+
+        old_ids = curr_members.map { |m| m['_id'] }
+        idx_old_members = curr_members.group_by { |m| m['host'] }.map_values(&:first)
+        idx_new_members = rs_members.group_by { |m| m['host'] }.map_values(&:first)
 
         ids = (0...256).to_a - old_ids
 
-        Chef::Log.info "after removing members, config = #{config}"
-        remaining_members = config['members'].count
+        new_members = idx_new_members.map { |h, m| idx_old_members[h].to_h.merge(m) }
```

Comment:
Is needed in case the key doesn't exists in the `idx_old_members` hash, so we default to `{}` so we can merge